File: backend/app/models/ui_test.py
"""UI Test Models.

This module defines the SQLAlchemy models for UI testing.
"""

from datetime import datetime, timezone
from typing import Optional, List
from sqlalchemy import Column, Integer, String, Boolean, Text, DateTime, ForeignKey, Enum
from sqlalchemy.orm import relationship
from app.database import Base
import enum


class Platform(str, enum.Enum):
    WEB = "web"
    APP = "app"


class UiTestStatus(str, enum.Enum):
    PENDING = "pending"
    RUNNING = "running"
    PASSED = "passed"
    FAILED = "failed"
    ERROR = "error"


class UiTestSuite(Base):
    """UI测试套件模型"""
    __tablename__ = "ui_test_suites"
    __allow_unmapped__ = True  # Allow legacy-style type annotations

    id: int = Column(Integer, primary_key=True, index=True)
    name: str = Column(String(100), nullable=False)
    description: Optional[str] = Column(Text)
    platform: Platform = Column(Enum(Platform), default=Platform.WEB)

    # 平台特定配置
    # Web: base_url, browser (chrome/firefox/safari)
    # APP: app_path, device_type
    config: Optional[str] = Column(Text)  # JSON格式

    created_by: Optional[str] = Column(String(50))
    created_at: datetime = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_by: Optional[str] = Column(String(50))
    updated_at: datetime = Column(
        DateTime,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc)
    )

    # Relationships
    cases: List["UiTestCase"] = relationship(
        "UiTestCase",
        back_populates="suite",
        cascade="all, delete-orphan"
    )

    def __repr__(self) -> str:
        return f"<UiTestSuite(id={self.id}, name={self.name!r}, platform={self.platform})>"


class UiTestCase(Base):
    """UI测试用例模型"""
    __tablename__ = "ui_test_cases"
    __allow_unmapped__ = True  # Allow legacy-style type annotations

    id: int = Column(Integer, primary_key=True, index=True)
    suite_id: int = Column(Integer, ForeignKey("ui_test_suites.id"), nullable=False, index=True)
    name: str = Column(String(200), nullable=False)
    description: Optional[str] = Column(Text)
    order: int = Column(Integer, default=0)
    is_enabled: bool = Column(Boolean, default=True)

    # Gherkin特征文件内容
    feature_content: Optional[str] = Column(Text)

    # 步骤定义 (JSON格式)
    # [{"keyword": "Given", "text": "打开登录页面", "action": "open_url", "params": {...}}]
    steps: Optional[str] = Column(Text)

    created_by: Optional[str] = Column(String(50))
    created_at: datetime = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_by: Optional[str] = Column(String(50))
    updated_at: datetime = Column(
        DateTime,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc)
    )

    # Relationships
    suite: Optional["UiTestSuite"] = relationship("UiTestSuite", back_populates="cases")
    executions: List["UiTestExecution"] = relationship(
        "UiTestExecution",
        back_populates="case",
        cascade="all, delete-orphan"
    )

    def __repr__(self) -> str:
        return f"<UiTestCase(id={self.id}, name={self.name!r})>"


class UiTestStepResult(Base):
    """UI测试步骤结果"""
    __tablename__ = "ui_test_step_results"
    __allow_unmapped__ = True  # Allow legacy-style type annotations

    id: int = Column(Integer, primary_key=True, index=True)
    execution_id: int = Column(Integer, ForeignKey("ui_test_executions.id"), nullable=False, index=True)

    step_order: int = Column(Integer)
    keyword: str = Column(String(20))  # Given/When/Then/And
    text: str = Column(Text)  # 步骤描述
    status: UiTestStatus = Column(Enum(UiTestStatus))
    error_message: Optional[str] = Column(Text)
    screenshot_path: Optional[str] = Column(String(500))
    duration_ms: Optional[int] = Column(Integer)

    # Relationships
    execution: Optional["UiTestExecution"] = relationship("UiTestExecution", back_populates="step_results")


class UiTestExecution(Base):
    """UI测试执行记录模型"""
    __tablename__ = "ui_test_executions"
    __allow_unmapped__ = True  # Allow legacy-style type annotations

    id: int = Column(Integer, primary_key=True, index=True)
    case_id: int = Column(Integer, ForeignKey("ui_test_cases.id"), nullable=False, index=True)
    batch_id: Optional[str] = Column(String(50), index=True)

    # 执行结果
    status: UiTestStatus = Column(Enum(UiTestStatus), default=UiTestStatus.PENDING)
    duration_ms: Optional[int] = Column(Integer)
    error_message: Optional[str] = Column(Text)

    # 附件
    screenshot_paths: Optional[str] = Column(Text)  # JSON数组
    video_path: Optional[str] = Column(String(500))
    log_path: Optional[str] = Column(String(500))

    executed_at: datetime = Column(DateTime, default=lambda: datetime.now(timezone.utc))

    # Relationships
    case: Optional["UiTestCase"] = relationship("UiTestCase", back_populates="executions")
    step_results: List["UiTestStepResult"] = relationship(
        "UiTestStepResult",
        back_populates="execution",
        cascade="all, delete-orphan"
    )

    def __repr__(self) -> str:
        return f"<UiTestExecution(id={self.id}, case_id={self.case_id}, status={self.status})>"

File: frontend/src/api/ui-test.ts
import axios from 'axios'

const api = axios.create({
  baseURL: '/api/v1',
  timeout: 60000 // UI 测试执行时间较长
})

// =====================
// Enum Constants
// =====================

// 平台类型
export const PLATFORMS = [
  { value: 'web', label: 'Web' },
  { value: 'app', label: 'App' },
] as const

// 浏览器类型
export const BROWSER_TYPES = [
  { value: 'chromium', label: 'Chromium' },
  { value: 'firefox', label: 'Firefox' },
  { value: 'webkit', label: 'WebKit' },
] as const

// 执行状态
export const EXECUTION_STATUSES = [
  { value: 'pending', label: '待执行' },
  { value: 'running', label: '执行中' },
  { value: 'passed', label: '通过' },
  { value: 'failed', label: '失败' },
  { value: 'error', label: '错误' },
] as const

// 步骤关键字
export const STEP_KEYWORDS = ['Given', 'When', 'Then', 'And', 'But'] as const

// =====================
// Type Definitions
// =====================

// UI 测试套件
export interface UiTestSuite {
  id: number
  name: string
  description?: string
  platform: 'web' | 'app'
  config?: UiSuiteConfig
  created_by?: string
  created_at: string
  updated_by?: string
  updated_at: string
  cases?: UiTestCase[]
}

// UI 套件配置
export interface UiSuiteConfig {
  base_url?: string
  browser?: 'chromium' | 'firefox' | 'webkit'
  headless?: boolean
  viewport?: { width: number; height: number }
  timeout?: number
  device_id?: string
  app_package?: string
}

// UI 测试用例
export interface UiTestCase {
  id: number
  suite_id: number
  name: string
  description?: string
  order: number
  is_enabled: boolean
  feature_content?: string
  steps?: UiTestStep[]
  created_by?: string
  created_at: string
  updated_by?: string
  updated_at: string
}

// UI 测试步骤
export interface UiTestStep {
  keyword: string
  text: string
  action: string
  params?: Record<string, any>
}

// UI 测试执行记录
export interface UiTestExecution {
  id: number
  case_id: number
  batch_id?: string
  status: string
  duration_ms?: number
  error_message?: string
  screenshot_paths?: string[]
  video_path?: string
  log_path?: string
  executed_at: string
  step_results?: UiTestStepResult[]
}

// UI 测试步骤结果
export interface UiTestStepResult {
  id: number
  execution_id: number
  step_order: number
  keyword: string
  text: string
  status: string
  error_message?: string
  screenshot_path?: string
  duration_ms?: number
}

// 执行结果
export interface ExecutionResult {
  success: boolean
  steps: Array<{
    order: number
    action: string
    success: boolean
    error: string | null
    duration_ms: number
    screenshot: string | null
  }>
  total_duration_ms: number
  error: string | null
  screenshots: string[]
}

// 套件列表响应
export interface UiTestSuiteListResponse {
  total: number
  items: UiTestSuite[]
}

// 执行列表响应
export interface UiTestExecutionListResponse {
  total: number
  items: UiTestExecution[]
}

// =====================
// API Methods
// =====================

export const uiTestApi = {
  // Suite operations
  getSuites: (skip = 0, limit = 100) =>
    api.get<UiTestSuiteListResponse>('/ui-tests/suites', {
      params: { skip, limit }
    }),

  getSuite: (id: number) =>
    api.get<UiTestSuite>(`/ui-tests/suites/${id}`),

  createSuite: (data: Partial<UiTestSuite>) =>
    api.post<UiTestSuite>('/ui-tests/suites', data),

  updateSuite: (id: number, data: Partial<UiTestSuite>) =>
    api.put<UiTestSuite>(`/ui-tests/suites/${id}`, data),

  deleteSuite: (id: number) =>
    api.delete(`/ui-tests/suites/${id}`),

  // Case operations
  getCases: (suiteId: number) =>
    api.get<UiTestCase[]>(`/ui-tests/suites/${suiteId}/cases`),

  getCase: (id: number) =>
    api.get<UiTestCase>(`/ui-tests/cases/${id}`),

  createCase: (data: Partial<UiTestCase>) =>
    api.post<UiTestCase>(`/ui-tests/suites/${data.suite_id}/cases`, data),

  updateCase: (id: number, data: Partial<UiTestCase>) =>
    api.put<UiTestCase>(`/ui-tests/cases/${id}`, data),

  deleteCase: (id: number) =>
    api.delete(`/ui-tests/cases/${id}`),

  // Execution operations
  executeCase: (id: number) =>
    api.post<UiTestExecution>(`/ui-tests/cases/${id}/execute`),

  executeSuite: (suiteId: number, caseIds?: number[]) =>
    api.post<ExecutionResult>(`/ui-tests/suites/${suiteId}/execute`, {
      suite_id: suiteId,
      case_ids: caseIds
    }),

  getExecutions: (caseId: number, skip = 0, limit = 20) =>
    api.get<UiTestExecutionListResponse>(`/ui-tests/cases/${caseId}/executions`, {
      params: { skip, limit }
    }),

  getExecution: (id: number) =>
    api.get<UiTestExecution>(`/ui-tests/executions/${id}`),
}

export default uiTestApi
